### PR TITLE
[Fury] Frothing Berserker rework + Juggernaut removal + Enrage

### PR DESCRIPTION
**Pull request by wisnoi**
Juggernaut no longer exists, and Frothing Berserker has been properly reworked and statistics implemented

Because of bad commit practice, Enrage has to be lumped here as well. Yikes.

### CODE REVIEW

File: src/parser/warrior/fury/modules/buffdebuff/Enrage.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        minor: 0.7,
+        average: 0.65,
+        major: 0.6,
+      },
+      style: 'percentage',
+    };
+  }
+
+  suggestions(when) {
+    const {
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/buffdebuff/Enrage.js]: I dont think you need this part here, all that gets done automatically if you do the below

suggestions(when) {
    when(this.suggestionThresholds)
        .addSuggestion((suggest, actual, recommended) => {

File: src/parser/warrior/fury/modules/buffdebuff/Enrage.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      this.damage += calculateEffectiveDamage(event, this.statTracker.currentMasteryPercentage);
+      this.totalDamage += event.amount;
+    }
+  }
+
+  get uptime() {
+    return this.selectedCombatant.getBuffUptime(SPELLS.ENRAGE.id) / this.owner.fightDuration;
+  }
+
+  get damageTotal () {
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/buffdebuff/Enrage.js]: This is a little nit picky, but you are calling this damageTotal but you arent actually outputing a total amount of damage, you are outputting the Damage Per Second increase. Might want to rename this to be more descriptive of what you are actually getting here.

File: src/parser/warrior/fury/modules/spells/MissedRampage.js [hunk 3]
```diff
@@ -2,57 +2,88 @@ import React from 'react';
...
+      };
+    }
   }
 
   suggestions(when) {
     when(this.suggestionThresholds).addSuggestion((suggest, actual, recommended) => {
       return suggest(
         <>
-          There were {actual} times when you cast another ability when you should have cast <SpellLink id={SPELLS.RAMPAGE.id} />.
+          There were {actual} times you casted a rage generating ability when you should have cast <SpellLink id={SPELLS.RAMPAGE.id} />.
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/spells/MissedRampage.js]: I would also include information on why they should have cast Rampage instead. This will help players that are new to the spec to understand why they are doing certain things or why it is a mistake to not do so.

File: src/parser/warrior/fury/modules/talents/FrothingBerserker.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    };
+
+    damage = 0;
+
+    constructor(...args) {
+        super(...args);
+
+        this.active = false;
+
+        if (this.selectedCombatant.hasTalent(SPELLS.FROTHING_BERSERKER_TALENT.id)) {
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/talents/FrothingBerserker.js]: You can shorten this to the following

this.active = this.selectedCombatant.hasTalent(SPELLS.FROTHING_BERSERKER_TALENT.id);
this.addEventListener(Events.damage.by(SELECTED_PLAYER, this.onPlayerDamage);

dont need the if statement for the event listener cause if its not active it wont run anyway.

File: src/parser/warrior/fury/modules/talents/FrothingBerserker.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            minor: 0.65,
+            average: 0.6,
+            major: 0.55,
+          },
+          style: 'percentage',
+        };
+      }
+
+    suggestions(when) {
+        const {
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/talents/FrothingBerserker.js]: Again, you dont need to do this, the info will get pulled from the suggestion thresholds if you do "when(this.suggestionThresholds)"

File: src/parser/warrior/fury/modules/talents/FrothingBerserker.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            isLessThan: {
+              minor,
+              average,
+              major,
+            },
+          } = this.suggestionThresholds;
+
+        when(true).isTrue().addSuggestion((suggest) => {
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/talents/FrothingBerserker.js]: Not sure about other people, but personally im not a huge fan of telling the player its a mistake to use the talent they picked unless one talent is vastly greater than the other. Im not going to tell you to change it cause i have no idea how close these talents actually are .... and its just my personal opinion as someone who believes players should play the way they want to play (if the talents are close). Feel free to leave it in though, I know something similar was added for Rune of Power for mages so its fine to have it. I would say just make sure you are only doing this when it is important and not for every talent that isnt the "Cookie cutter best build"

File: src/parser/warrior/fury/modules/spells/MissedRampage.js [hunk 7]
```diff
@@ -2,57 +2,88 @@ import React from 'react';
...
 
-const GENERATORS = [
-  SPELLS.RAGING_BLOW.id,
-  SPELLS.BLOODTHIRST.id,
-  SPELLS.EXECUTE.id,
+/*  Example log:
+ *  https://www.warcraftlogs.com/reports/KhynM7v96cZkTBdg#fight=6&type=damage-done&source=78
+ */
+
+const rageGenerators = [
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/spells/MissedRampage.js]: i believe the standard for constants declared up top is to use all caps, so RAGE_GENERATORS instead of rageGenerators. Not sure if that is a hard rule or not, just what ive seen in all the modules ive looked at.

File: src/parser/warrior/fury/modules/buffdebuff/Enrage.js (hunk 1 above)
Comment (matardarix) [reply src/parser/warrior/fury/modules/buffdebuff/Enrage.js]: works only if you add `actual: this.enrageUptime, ` inside the suggestion thresholds tho

Comment (Sharrq) [reply src/parser/warrior/fury/modules/buffdebuff/Enrage.js]: yea i missed that. Here is an example from one of the mage modules

  get utilSuggestionThresholds() {
    return {
      actual: this.utilPercent,
      isLessThan: {
        minor: 0.95,
        average: 0.90,
        major: 0.80,
      },
      style: 'percentage',
    };
  }

File: src/parser/warrior/fury/modules/talents/FrothingBerserker.js (hunk 6 above)
Comment (wisnoi) [reply src/parser/warrior/fury/modules/talents/FrothingBerserker.js]: There is absolutely unequivocally no situation in which this talent is an increase in DPS, ever. It's a non-choice. It's not so much cookie cutter as it has "gimping yourself for less DPS and clunkier gameplay". I'd love to see frothing actually be good, but it really isn't. :/

File: src/parser/warrior/fury/modules/buffdebuff/Enrage.js (hunk 1 above)
Comment (Sharrq) [reply src/parser/warrior/fury/modules/buffdebuff/Enrage.js]: basically change this

```
suggestions(when) {
    const {
      isLessThan: {
        minor,
        average,
        major,
      },
    } = this.suggestionThresholds;
     when(this.enrageUptime).isLessThan(minor)
``

to this
```
suggestions(when) {
    when(this.suggestionThresholds)
```

and add `actual: this.enrageUptime,` above `isLessThan: {` in the suggestionThresholds

File: src/parser/warrior/fury/modules/talents/FrothingBerserker.js (hunk 6 above)
Comment (Sharrq) [reply src/parser/warrior/fury/modules/talents/FrothingBerserker.js]: Thats fine. Then absolutely leave it in.

File: src/parser/warrior/fury/modules/talents/FrothingBerserker.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            return suggest(
+              <>
+              You are currently using <SpellLink id={SPELLS.FROTHING_BERSERKER_TALENT.id} />. It is highly recommended to talent into <SpellLink id={SPELLS.CARNAGE_TALENT.id} />, as there are currently no situations in which  <SpellLink id={SPELLS.FROTHING_BERSERKER_TALENT.id} /> outperforms <SpellLink id={SPELLS.CARNAGE_TALENT.id} />.
+              </>
+            )
+            .icon(SPELLS.CARNAGE_TALENT.icon)
+            .staticImportance(SUGGESTION_IMPORTANCE.REGULAR);
+        });
+    
+        when(this.uptimePercent).isLessThan(0.65)
```
Comment (Sharrq) [inline src/parser/warrior/fury/modules/talents/FrothingBerserker.js]: same as the Enrage Module, this can be `when(this.suggestionThresholds)` instead
